encoding/template/run_length_encode.go

```diff
@@ -0,0 +1,80 @@
+package template
+
+import (
+	"encoding/binary"
+	"io"
+
+	"github.com/mauricelam/genny/generic"
+	"github.com/xichen2020/eventdb/x/bytes"
+)
+
+// GenericValue is a generic type.
+type GenericValue generic.Type
+
+// ValueMarshalFn reads a GenericValue from an `io.Reader`.
+type ValueMarshalFn func(writer io.Writer, value GenericValue) error
+
+// ForwardValueIterator allows iterating over a stream of GenericValue.
+type ForwardValueIterator interface {
+	generic.Type
+
+	io.Closer
+	Next() bool
+	Err() error
+	Current() GenericValue
+}
+
+// runLengthEncodeValue run length encodes a stream of GenericValue.
+func runLengthEncodeValue(
+	writer io.Writer,
+	extBuf *[]byte, // extBuf is an external byte buffer for memory re-use.
+	marshalFn ValueMarshalFn,
+	valuesIt ForwardValueIterator,
+) error {
+	// Ensure that our buffer size is large enough to handle varint ops.
+	*extBuf = bytes.EnsureBufferSize(*extBuf, binary.MaxVarintLen64, bytes.DontCopyData)
+
+	var (
+		last        *GenericValue
```

Comment:
Can you use `GenericValue` instead of `*GenericValue`? I'd think if it's a type alias of `interface{}` using `GenericValue` should work?

If you want to check for first value, you can define a separator `firstTime` var for clarity, initialize it to `true,` and set it to `false` during iteration.

Reply:
Yea, I was doing it to catch the first value. 👍